### PR TITLE
add test for oss-fuzz

### PR DESCRIPTION
**Pull request by sebpop**
Move the fuzz checker code upstream as requested in this review https://github.com/google/oss-fuzz/pull/1694

### CODE REVIEW

File: Makefile.in [hunk 1]
```diff
@@ -85,9 +85,18 @@ OBJS = $(OBJC)
 
 PIC_OBJS = $(PIC_OBJC)
 
+# By default, use our own standalone_fuzz_target_runner.
+# This runner does no fuzzing, but simply executes the inputs
+# provided via parameters.
+# Run e.g. "make all LIB_FUZZING_ENGINE=/path/to/libFuzzer.a"
+# to link the fuzzer(s) against a real fuzzing engine.
+#
+# OSS-Fuzz will define it's own value for LIB_FUZZING_ENGINE.
```
Comment (mtl1979) [inline Makefile.in]: Should be "its", not "it's"

File: test/fuzz/compress_fuzzer.c [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "zlib-ng.h"
+
+static Bytef cbuf[256 * 1024] = { 0 };
+static Bytef ubuf[256 * 1024] = { 0 };
```
Comment (mtl1979) [inline test/fuzz/compress_fuzzer.c]: "Bytef" is discouraged in new code, should use standard types instead.

File: test/fuzz/compress_fuzzer.c [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  uLongf cbuf_len = (uLongf)sizeof(cbuf);
+  uLongf ubuf_len = (uLongf)sizeof(ubuf);
```
Comment (mtl1979) [inline test/fuzz/compress_fuzzer.c]: "uLongf" is discouraged in new code, should use standard types instead.

File: Makefile.in (hunk 1 above)
Comment (sebpop) [reply Makefile.in]: Right.  I copied that comment from the example at https://github.com/google/oss-fuzz/blob/master/projects/example/my-api-repo/Makefile
I will fix it in my patch and send out a patch to fix the example.

File: test/fuzz/standalone_fuzz_target_runner.c [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    size_t len = ftell(f);
+    fseek(f, 0, SEEK_SET);
+    unsigned char *buf = (unsigned char*)malloc(len);
+    size_t n_read = fread(buf, 1, len, f);
+    assert(n_read == len);
+    LLVMFuzzerTestOneInput(buf, len);
+    free(buf);
+    fprintf(stderr, "Done:    %s: (%zd bytes)\n", argv[i], n_read);
+  }
+}
```
Comment (mtl1979) [inline test/fuzz/standalone_fuzz_target_runner.c]: Missing return statement.

**Comment by Dead2** [comment]
Currently this fuzzer only tests with the default compression level (6), meaning either deflate_medium or _slow, depending on compilation choices. Maybe it would be better to get each of the compression algorithms tested equally? That would require changing to using the compress2 function, and selecting a compression level.

Have a look at configuration_table in deflate.c for example. We have deflate_quick, _fast, _medium and _slow being used at different compression levels. 

How about the fuzzer takes a buffer and compresses it with levels 1, 3, 6, 7 and verifies each of the returned compressed buffers. (lvl 6 could be changed to lvl 4 to increase speed, but since lvl 6 is the default level, that might be an even more compelling reason not to.)


Btw, from a security standpoint, testing broken compressed streams through the decompressor would likely be very valuable. This is where a corrupt .png/.pdf etc file would be able to cause havoc on the unsuspecting users machine.

**Comment by sebpop** [comment]
These are very good points: I will add code to the compress_fuzzer to check compression at different levels. 
I will add a decompress_fuzzer that starts decompressing the inputs of the fuzzer, then it will compress that and finally match that the output of decompress+compress matches the original input.

**Comment by sebpop** [comment]
> then it will compress that and finally match that the output of decompress+compress matches the original input.

This last part for the decompress checker does not seem to be feasible: there is no unique compression form.

**Comment by Dead2** [comment]
The decompression tester might need a semi-valid header to actually get through to the decompression code more often, and not just fail during the init code (I did not verify that it does bail out during init, but I suspect it might).
Perhaps by testing both without a header and with a prepended header that just specifies the most common mode of using deflate with no preset dictionary.

Btw, maybe the bugfixes should go in a separate PR, so we don't clutter up this PR with comments regarding those.

**Comment by sebpop** [comment]
Now that the oss-fuzzers are free of bugs and that the patches have stabilized, I think it is time to think about merging the patch series.
@Dead2 : could you please let me know how you would like to proceed?
I can rework the patches following your guidelines and create separate pull requests for each fixed bug.

**Comment by Dead2** [comment]
@sebpop Separate pull requests for each fixed bug is great, just like I see you have started doing. 👍 
Also a separate PR for the oss-fuzz support would be great.
I am a bit unsure about the attribution requirements from oss-fuzz though, didn't they require some text in the commit message?

@mtl1979 Please +1 the PRs you approve of if you can, I value your feedback.
Any others doing the same would be great. The more eyes reviewing zlib-ng changes, the better both for me and for the projects potential inclusion in distros.  :)

@sebpop Great work on this! If you ever visit (east) Norway, I'll buy you a beer or something :)

**Comment by mtl1979** [comment]
@Dead2 Already started going through the ones that seem ready for merging... Some of the changes did seem still work-in-progress so I'm waiting for @sebpop to finish them. 

I'm currently quite busy with work so I can't go through the modifications every day, but I try to go through them when I have spare time.

**Comment by Dead2** [comment]
@mtl1979 Great, I am not rushing anything here. But knowing you intend to review, I'll hold off on merging until I hear your feedback on them. Oh, and that beer offer goes for you too :)

**Comment by mtl1979** [comment]
@Dead2 Everybody knows I don't drink beer...

**Comment by Dead2** [comment]
@mtl1979 Me neither ;)

**Comment by mtl1979** [comment]
@Dead2 I did empty my last bottle of alcohol about 1 year after I stopped working for local recycling center... After that I have not even tasted alcohol.

**Comment by sebpop** [comment]
> Also a separate PR for the oss-fuzz support would be great.

I just opened bug #204.

> I am a bit unsure about the attribution requirements from oss-fuzz though, didn't they require some text in the commit message?

For the bugs with an oss-fuzz ticket, I added that reference in the subject of the patch fixing it.
If you want I can also add the reference in the commit message.

> the projects potential inclusion in distros

I'd like to get zlib-ng in chromium.org and Android.  oss-fuzzer is going to help stabilize the code, and the big performance difference between zlib-ng and zlib.org already provides the incentive to make the switch.

> If you ever visit (east) Norway, I'll buy you a beer or something :)

I like to keep my brain clear to work on code, I haven't got a beer from the time I was in Liverpool.
Hot chocolate or a tea would work ;-)

**Comment by sebpop** [comment]
I am closing this old pull request.  The squashed patch to add the fuzzers in a single patch is #206.

**Pull request closed without merging**